### PR TITLE
add update api

### PR DESCRIPTION
**Pull request by fredchenbj**
Signed-off-by: fredchenbj <cfworking@163.com>

## What have you changed? (mandatory)
Use rocksdb `merge`  to implement tikv's update api. I add update logic in tikv-server layer and add a new merge operator `update_merge` to present a table-style colomn's update.
Other Points: 
- to compile this pr, need to replace kvproto with pr(https://github.com/pingcap/kvproto/pull/342)
- and maybe fix this compile error of rust-rocksdb(https://github.com/pingcap/rust-rocksdb/issues/264)

## What are the type of the changes? (mandatory)
new feature

## How has this PR been tested? (mandatory)
manual tests

## Does this PR affect documentation (docs) update? (mandatory)
no

## Does this PR affect tidb-ansible update? (mandatory)
no

## Refer to a related PR or issue link (optional)
https://github.com/pingcap/kvproto/pull/342

### CODE REVIEW

**Comment by siddontang** [comment]
Very cool @fredchenbj

**Comment by siddontang** [comment]
We should notice that this is conflicted with Titan.

**Comment by siddontang** [comment]
PTAL @huachaohuang @DorianZheng

**Comment by ice1000** [comment]
Please sign-off all your commits, thanks!

**Comment by huachaohuang** [comment]
Thanks, but I think this API is not general-purpose, the merge logic is bound to the specific business.

File: src/config.rs [hunk 1]
```diff
@@ -1275,6 +1278,138 @@ pub fn check_and_persist_critical_config(config: &TiKvConfig) -> Result<(), Stri
...
 }
 
+/**
+|total_size|key1_size|key1|value1_size|value1|key2_size|key2|value2_size|value2|...
+all size is four bytes.
+existing_value : existing_val
+operand_list   : operands
+new_value      : result
+Note: little endian Coding
+*/
```
Comment (MyonKeminta) [inline src/config.rs]: Please consider using rustdoc style here, like:
```rustdoc
/// Comments...
/// ...
```

File: src/config.rs [hunk 2]
```diff
@@ -1275,6 +1278,138 @@ pub fn check_and_persist_critical_config(config: &TiKvConfig) -> Result<(), Stri
...
+    }
+
+    if let Some(existing_value) = existing_val {
+        let mut begin = 0;
+        let exist_total_size = existing_value.len();
+
+        begin += 4 ;
+
+        while begin < exist_total_size {
+            //decode key size
```
Comment (MyonKeminta) [inline src/config.rs]: Please place a space between `//` and following text, like:
```rust
// Decode key size.
```
The same to other comments here

File: src/raftstore/store/fsm/apply.rs [hunk 3]
```diff
@@ -1141,6 +1142,51 @@ impl ApplyDelegate {
         Ok(resp)
     }
 
+    // handle update
```
Comment (MyonKeminta) [inline src/raftstore/store/fsm/apply.rs]:
```suggestion
    /// Handles update
```

File: src/raftstore/store/fsm/apply.rs [hunk 4]
```diff
@@ -1141,6 +1142,51 @@ impl ApplyDelegate {
...
+                    escape(&key),
+                    escape(value),
+                    e
+                );
+            });
+        }
+        Ok(resp)
+    }
+
+
```
Comment (MyonKeminta) [inline src/raftstore/store/fsm/apply.rs]: Here is a redundant empty line. You can run `cargo fmt` to fix it.

**Comment by MyonKeminta** [comment]
By the way, please sign-off your commits. Thanks!

**Comment by fredchenbj** [comment]
> By the way, please sign-off your commits. Thanks!

@MyonKeminta Thanks, I will signed off new commits. But do I need signed off the commits that I had committed? Is there method to do this which I didn't find?

**Comment by MyonKeminta** [comment]
Oh, sorry. It seems I shouldn't care code style now since it needs discussion.
Could you please try to give an example to show why this is useful?

By the way, each commit should be signed off. For previously pushed commits, you may use git rebase, edit the commit messages and finally push -f.

**Comment by fredchenbj** [comment]
@MyonKeminta I have signed off the commits. And I would change the code to make it more general for other situations. Meanwhile, I will do more explanation about this pr after. Thanks.

**Pull request closed without merging**